python/ray/dag/dag_node.py

```diff
@@ -58,12 +60,41 @@ def __init__(
         self._bound_other_args_to_resolve: Optional[Dict[str, Any]] = (
             other_args_to_resolve or {}
         )
+
+        # The list of nodes that use this DAG node as input.
+        self._downstream_nodes: List["DAGNode"] = []
+        # The list of nodes that this DAG node uses as input.
+        self._upstream_nodes: List["DAGNode"] = self._prepare_upstream_nodes()
+
         # UUID that is not changed over copies of this node.
         self._stable_uuid = uuid.uuid4().hex
         # Cached values from last call to execute()
         self.cache_from_last_execute = {}
 
         self._type_hint: Optional[ChannelOutputType] = ChannelOutputType()
+        self.is_output_node = False
+
+    def _prepare_upstream_nodes(self) -> List["DAGNode"]:
+        """
+        Retrieve upstream nodes and update their downstream dependencies.
+
+        TODO (kevin85421): Currently, the upstream nodes and downstream nodes have
+        circular references. Therefore, it relies on the garbage collector to clean
+        them up.
```

Comment:
What's the consequence of this? Does it have memory leak or any side effects?

Reply:
It will not have memory leaks. Python's garbage collector can recycle it. 

> any side effects?

https://github.com/ray-project/ray/pull/47261#issuecomment-2306118753

Reply:
Yeah, I guess my question was what was the issue and the TODO exactly.

Reply:
Updated: https://github.com/ray-project/ray/pull/47261/commits/cd35b0be0947b8f0d3e5471cedcded2c76c47640